marker_api/src/diagnostic.rs

```diff
@@ -148,60 +219,167 @@ impl<'ast> DiagnosticBuilder<'ast> {
         span: &Span<'ast>,
         suggestion: impl ToString,
         app: Applicability,
-    ) {
-        self.parts.push(DiagnosticPart::Suggestion {
-            msg: msg.to_string(),
-            span: span.clone(),
-            sugg: suggestion.to_string(),
-            app,
-        });
+    ) -> &mut Self {
+        if self.emit_lint {
+            self.parts.push(DiagnosticPart::Suggestion {
+                msg: msg.to_string(),
+                span: span.clone(),
+                sugg: suggestion.to_string(),
+                app,
+            });
+        }
+        self
+    }
+
+    /// This function takes a closure, that is only executed, if the created
+    /// diagnostic is actually emitted in the end. This is useful for crafting
+    /// suggestions. Having them in a conditional closure will speedup the
+    /// linting process if the lint is allowed at a given location.
+    ///
+    /// ```
+    /// # use marker_api::prelude::*;
+    /// # marker_api::declare_lint!(
+    /// #     /// Dummy
+    /// #     LINT,
+    /// #     Warn,
+    /// # );
+    /// # fn value_provider<'ast>(cx: &AstContext<'ast>, node: ExprKind<'ast>) {
+    ///     cx.emit_lint(LINT, node, "<lint message>").decorate(|diag| {
+    ///         // This closure is only called, if the lint is enabled. Here you
+    ///         // can create a beautiful help message.
+    ///         diag.help("<text>");
+    ///     });
+    /// # }
+    /// ```
+    pub fn decorate<F>(&mut self, decorate: F) -> &mut Self
+    where
+        F: FnOnce(&mut DiagnosticBuilder<'ast>),
+    {
+        if self.emit_lint {
+            decorate(self);
+        }
+        self
     }
 
     pub(crate) fn emit<'builder>(&'builder self, cx: &AstContext<'ast>) {
-        let parts: Vec<_> = self.parts.iter().map(DiagnosticPart::to_ffi_part).collect();
-        let diag = Diagnostic {
-            lint: self.lint,
-            msg: self.msg.as_str().into(),
-            node: self.node,
-            span: &self.span,
-            parts: parts.as_slice().into(),
-        };
-        cx.emit_diagnostic(&diag);
+        if self.emit_lint {
+            let parts: Vec<_> = self.parts.iter().map(DiagnosticPart::to_ffi_part).collect();
+            let span = self
+                .span
+                .as_ref()
+                .expect("always `Some`, if `DiagnosticBuilder::emit_lint` is true");
```

Comment:
I'd make it this way:
```rust
struct DiagnosticBuilder {
    imp: Option<DiagnosticBuilderImp>
}

struct DiagnosticBuilderImp {
    // all fields are here
}
```
If the `imp` is `None` it means the lint shouldn't be emitted. It forces you at type level to write:
```rust
if let Some(diag) = &mut self.imp {
    // calculate the diagnostic here
}
```
When with `emit_lint` bool every field should have a trivially-constructible default value, and you also need to remember to put an `if self.emit_lint {}`.

Reply:
I'm considering this solution or having everything lazy, how you proposed it here https://github.com/rust-marker/marker/pull/245#discussion_r1328026700

> Eagerly setting message for the allowed lint may also impact performance.
> I wonder if the API could then look like this which forces all users to calculate the diagnostics lazily.
> 
> ```rust
> // Simple case
> cx.emit_lint(LINT, node, |_| "Simple message");
> 
> // For elaborate case
> cx.emit_lint(LINT, node, |diag| diag.message("message").help("help"));
> ```
> 
> To implement this the closure needs to have a return type that implements some trait e.g. `Diagnostic`, where everything that implements `Into<String>` also implements `Diagnostic`, and the `DiagnosticBuilder` too.
> 
> But otherwise, the API in this form is also workable. I don't have a strong preference

I understood your comments in the feedback issue, that you found the closure rather confusing. In which case, the first option would be better. :thinking:

Reply:
I don't see how putting everything under an `Option` in `DiagnosticBuilder` influences the `emit_lint()` method signature
So what signatures are you comparing against?

Reply:
Well, if the closure taking the diagnostic builder is only executed, if the lint is emitted. We would only instantiate the diagnostic builder, to call the closure, meaning that the fields don't need to be conditional.

Reply:
Nope, my suggestion was such that the `emit_lint` method would return `()`. I guess you are right, in case the closure is passed as a parameter to `emit_lint` then nothing optional is needed in the builder

Reply:
Okay, the suggested API seems quite ergonomic. I'll probably think a bit more about this, before I move forward in either direction. Thank you for the input, I really appreciate it!

Reply:
I like having a trait for the last parameter of `emit_lint`, I've done some testing, but sadly got stuck. Do you maybe have any ideas on how to fix it? This is the trait and the implementations

```rs
pub trait IntoDiagnostic<'ast> {
    fn build_diagnostic(self, builder: DiagnosticBuilder<'ast>);
}

impl<'ast> IntoDiagnostic<'ast> for &str {
    fn build_diagnostic(self, mut builder: DiagnosticBuilder<'ast>) {
        builder.set_main_message(self);
    }
}
impl<'ast> IntoDiagnostic<'ast> for String {
    fn build_diagnostic(self, mut builder: DiagnosticBuilder<'ast>) {
        builder.set_main_message(self);
    }
}
impl<'ast, F: FnOnce(DiagnosticBuilder<'ast>)> IntoDiagnostic<'ast> for F {
    fn build_diagnostic(self, builder: DiagnosticBuilder<'ast>) {
        self(builder);
    }
}
```

With text it works just fine, but when I try to use a closure, I get the following error:

```
error[E0282]: type annotations needed
   --> marker_uilints/src/lib.rs:201:47
    |
201 |         cx.emit_lint(PRINT_EVERY_EXPR, expr, |mut diag| {
    |                                               ^^^^^^^^
202 |             diag.set_main_message("expr")
    |             ---- type must be known at this point
    |
help: consider giving this closure parameter an explicit type
    |
201 |         cx.emit_lint(PRINT_EVERY_EXPR, expr, |mut diag: /* Type */| {
    |
```

The type requirement makes sense to me, at the same time I would like this to be optional. Typing `: DiagnosticBuilder<'_>` seems less ergonomic. I assume this happens because `diag` is mutable. I've also tried having the type be `&mut DiagnosticBuilder<'_>` but that also requires the type annotation :thinking: 

Currently, I can think of three solutions:
1. Require the type annotation
2. Use internal mutability
3. I'm missing something :thinking: 

I pushed the last (broken) state with `&mut DiagnosticBuilder<'_>` as a parameter.

Reply:
@Veetaha Do you maybe have an idea how to solve the problem?

I couldn't find a solution. Now, I'm thinking of having two methods:
1. `Context::emit_lint(<lint>, <node>, <message>)`
2. `Context::emit_lint_and_decorate(<lint>, <node>, <message>, <decoration closure>)`

Reply:
Yeah, that's a tough problem. I know, for example, [reqwest_middleware::RequestInitialiser](https://docs.rs/reqwest-middleware/latest/reqwest_middleware/trait.RequestInitialiser.html) suffers from the same. This inference problem isn't specific to mutability, I tried `&` and by-value, but it still requires a type annotation.

I'll try to think more about this.

Reply:
Thank you!

I've now implemented the second option `emit_lint` and `emit_lint_and_decorate` for comparison. It seems to work fairly well. It's a little less flexible but easier to use IMO. You can checkout the latest commit for that solution or [2204702](https://github.com/rust-marker/marker/pull/245/commits/2204702ec888d381439dbdab333a6d95d6d14639) for `IntoDiagnostic`

I'll fix the CI later, for now, it's just a showcase of the two options :)

Reply:
> Maybe the diagnostic builder shouldn't expose the set_main_message method?

Yeah, I think that method should be removed. I wanted to reduce the number of arguments, but that one was a bit too aggressive.

> I was originally commenting that constructing the message could be non-trivial and should also be in closure, but maybe that's not actually a problem?

I liked the theoretical freedom of your proposed solution. The main messages themselves are usually very trivial. Only a tiny fraction of Clippy's have a dynamic main message, usually it's just a string literal.

> This will only matter when the lint is disabled, and there should generally be quite small amount of `allows` in code.

This highly depends on the lint authors and users of Marker. Currently, I only have Clippy as a reference, where most lints are allow-by-default. It can be expected most Marker lints will be warn-by-default. But if they have a higher false positive (FP) rate, they're also likely to be allowed more.  

> So I guess your original API is fine.

Okay, thank you for all the input! We can also adapt this later if needed.

> Regarding the difference between `emit_lint_and_decordate()` and `emit_lint().decordate()`. The latter looks cleaner but may be harder to discover.

We can always create a lint to suggest the usage of `.decorate()` :wink: